### PR TITLE
WIP: Sketch for improving Engine

### PR DESCRIPTION
**Pull request by pbrisbin**
/cc @codeclimate/review

This is as sketch. It's built on top of pb-container-log. If we think this is a
good direction, I'll wait for that to merge and open a new PR from this branch
after rebasing, adding/fixing tests etc.

---

New responsibility breakdown:

**Container**: launch a docker container, give its stdout to the caller,
capture its stderr, monitor its process, handle timeouts, call life-cycle
events on the container-log given.

**Engine::ContainerLog**: perform the statsd calls on container life-cycle
events and then forward those events to the log its been given; sort of a
ContainerLog middleware stack idea.

**Engine**: launch and run a Container for the engine, specify how to shuttle
null-delimited stdout to the formatter, compose its container-log with the one
given.

Benefits:

- Engine is smaller, more focused
- Container is reusable for Clone in builder
- Container is easier to reason about (i.e. the complex timeout logic is
  clearer IMO)

Notable change:

Rather than spawning `docker kill` I'm using Ruby's `Process.kill`, the two
should be equivalent and this was more convenient with the way the code is now
(e.g. no need to track container-name).

### CODE REVIEW

File: lib/cc/analyzer/container.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+      def on_output(delimeter = "\n", &block)
+        @output_delimeter = delimeter
+        @on_output = block
+      end
+
+      def run(options)
+        @log.started(@image)
+
+        pid, _, out, err = POSIX::Spawn.popen4(*docker_run_command)
```
Comment (pbrisbin) [inline lib/cc/analyzer/container.rb]: I forgot to implement `docker_run_command` but it'll be `docker run` + default options + `options`

File: lib/cc/analyzer/container.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    class Container
+      TIMEOUT = 15 * 60 # 15m
+
+      def initialize(image, command = nil, log = nil)
+        @image = image
+        @command = command
+        @log = log || NullContainerLog.new
+
+        @output_delimeter = "\n"
+        @on_output = ->(*)
```
Comment (pbrisbin) [inline lib/cc/analyzer/container.rb]: Should be `->(*) { }` basically a no-op default.

File: lib/cc/analyzer/container.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      def run(options)
+        @log.started(@image)
+
+        pid, _, out, err = POSIX::Spawn.popen4(*docker_run_command)
+
+        t_out = read_stdout(out)
+        t_err = read_stderr(err)
+        t_timeout = timeout_thread(pid)
+
+        _, status = Process.waitpid2(pid)
```
Comment (pbrisbin) [inline lib/cc/analyzer/container.rb]: Forgot to call `@log.finished(status, @stderr_io.string)`

File: lib/cc/analyzer/engine.rb [hunk 4]
```diff
@@ -16,83 +17,28 @@ def initialize(name, metadata, code_path, config, label)
...
-        container_log.started(@metadata["image"])
-
-        timed_out = false
-        pid, _, out, err = POSIX::Spawn.popen4(*docker_run_command)
-        Analyzer.statsd.increment("cli.engines.started")
+      def run(stdout_io, container_log = NullContainerLog.new)
+        container = Container.new(
+          @metadata["image"],
+          @metadata["command"],
+          ContainerLogLog.new(name, container_log)
```
Comment (GordonDiggs) [inline lib/cc/analyzer/engine.rb]: `ContainerLogLog`?

File: lib/cc/analyzer/engine.rb [hunk 5]
```diff
@@ -113,21 +61,48 @@ def config_file
...
-      end
-
       def output_filter
         @output_filter ||= EngineOutputFilter.new(@config)
       end
 
-      CommandFailure = Class.new(StandardError)
-      EngineFailure = Class.new(StandardError)
-      EngineTimeout = Class.new(StandardError)
+      class ContainerLog
```
Comment (pbrisbin) [inline lib/cc/analyzer/engine.rb]: FYI: I'll move this to its own file in the real PR.

File: lib/cc/analyzer/engine.rb (hunk 4 above)
Comment (pbrisbin) [inline lib/cc/analyzer/engine.rb]: Heh, would clearly catch stuff like this when I do the real PR with tests.

**Comment by pbrisbin** [comment]
FYI: because of the rebase on the base branch, this is now showing some differences that aren't really intended. Sorry.

**Comment by pbrisbin** [comment]
Spoke with Gordon IRL -- we think this is promising enough to move forward. I'll open a new PR with tests and things shortly.

**Pull request closed without merging**